cpp/ql/src/Likely Bugs/OO/UnsafeUseOfThis.ql

```diff
@@ -0,0 +1,288 @@
+  flowsToSink(succ, _, sinkClass) and
+  (
+    irBbPostDominates(succ.getBlock(), instr.getBlock()) and
+    (
+      (
+        succ.(CopyInstruction).getSourceValue() = instr or
+        succ.(CheckedConvertOrNullInstruction).getUnary() = instr or
+        succ.(ChiInstruction).getTotal() = instr or
+        succ.(ConvertInstruction).getUnary() = instr
+      )
+      or
+      succ.(InheritanceConversionInstruction).getUnary() = instr
+    )
+    or
+    flowThroughCallable(instr, succ, sinkClass)
+  )
+}
+
+/**
+ * Holds if
+ * - `instr` is an argument (or argument indirection) to a call, and
+ * - `succ` is the corresponding initialization instruction in the call target, and
+ * - there exists an initialization of a `this` pointer of type `sourceClass` that flows to `instr`.
+ */
+predicate flowThroughCallableFwd(Instruction instr, Instruction succ, Class sourceClass) {
+  flowsFromSource(_, instr, sourceClass) and
+  (
+    // Flow from an argument to a parameter
+    exists(CallInstruction call, InitializeParameterInstruction init | init = succ |
+      getPositionalArgumentInitParam(call, instr, init, call.getStaticCallTarget())
+      or
+      getThisArgumentInitParam(call, instr, init, call.getStaticCallTarget())
+    )
+    or
+    // Flow from argument indirection to parameter indirection
+    exists(
+      CallInstruction call, ReadSideEffectInstruction read, InitializeIndirectionInstruction init
+    |
+      init = succ and
+      init.getEnclosingFunction() = call.getStaticCallTarget() and
+      read.getPrimaryInstruction() = call and
+      read.getSideEffectOperand().getAnyDef() = instr
+    |
+      exists(int n |
+        read.getIndex() = n and
+        init.getParameter().getIndex() = unbind(n)
+      )
+      or
+      call.getThisArgument() = instr and
+      init.getIRVariable() instanceof IRThisVariable
+    )
+  )
+}
+
+/**
+ * Holds if there exists an initialization of a `this` pointer of type `sourceClass` that flows
+ * to `instr`, and `instr` flows to `succ` and `succ`.
+ */
+predicate successorFwd(Instruction instr, Instruction succ, Class sourceClass) {
+  flowsFromSource(_, instr, sourceClass) and
+  (
+    irBbPostDominates(succ.getBlock(), instr.getBlock()) and
+    (
+      (
+        succ.(CopyInstruction).getSourceValue() = instr or
+        succ.(CheckedConvertOrNullInstruction).getUnary() = instr or
+        succ.(ChiInstruction).getTotal() = instr or
+        succ.(ConvertInstruction).getUnary() = instr
+      )
+      or
+      succ.(InheritanceConversionInstruction).getUnary() = instr
+    )
+    or
+    flowThroughCallableFwd(instr, succ, sourceClass)
+  )
+}
+
+/**
+ * Holds if `instr` is in the path from an initialization of a `this` pointer in a subclass, to a use
+ * of the `this` pointer in a baseclass.
+ */
+predicate isInPath(Instruction instr) {
+  exists(Class sourceClass, Class sinkClass |
+    flowsFromSource(_, instr, sourceClass) and
+    flowsToSink(instr, _, sinkClass) and
```

Comment:
It seems right to check, as you do here, that there's flow both forward and backwards so we avoid producing too many nodes for path explanations. But is it definitely necessary to redo the analysis from scratch in both directions, with a set of very similar-looking predicates (with and without `Fwd`)?

Can't we define `isInPath` as a recursive predicate with a base case of `isSource` and a recursive case of `isInPath` joined with `successor`? Then there would be no `Fwd` predicates.

Another option is to do what the data-flow library does and use a backward-then-forward scan as an optimisation by making the backward scan unary. The initial backward scan just tells us which nodes can reach a sink without telling us which sink it is. Then the main analysis is to do a forward scan from all sources that can reach a sink, visiting only intermediate nodes that can reach a sink. Then `isInPath` is any node that was reached in the forward scan.

Reply:
As I've already taken inspiration from the dataflow library on how to get the proper join order for parameter/arguments flow, it seemed like a good idea to follow the backward-then-forward optimization. Hopefully cce682c74715901b29000339ea6117e24c833e1e contains what you had in mind.